ihmc-perception/src/main/java/us/ihmc/perception/ImageSensorPublishThread.java

```diff
@@ -24,19 +28,51 @@ public ImageSensorPublishThread(ROS2Node ros2Node, ImageSensor sensorToPublish,
       publisher = new RawImagePublisher(ros2Node);
    }
 
+   public ImageSensorPublishThread(ROS2Node ros2Node,
+                                   ImageSensor sensorToPublish,
+                                   Map<Integer, ROS2Topic<? extends Packet<?>>> imageKeyToTopicMap,
+                                   double publishFrequencyLimit)
+   {
+      this(ros2Node, sensorToPublish, imageKeyToTopicMap);
+      setFrequencyLimit(publishFrequencyLimit);
+   }
+
+   @Override
+   public ImageSensorPublishThread setFrequencyLimit(double publishFrequencyLimit)
+   {
+      super.setFrequencyLimit(publishFrequencyLimit);
+      publishingIsThrottled = publishFrequencyLimit > 0.0;
+      return this;
+   }
+
    @Override
    protected void runTask()
    {
       try
-      {  // Wait for images to be grabbed
-         imageSensor.waitForGrab();
+      {  // If publisher frequency is not set, run at sensor's frequency
+         if (!publishingIsThrottled)
+            imageSensor.waitForGrab();
 
          for (Entry<Integer, ROS2Topic<? extends Packet<?>>> imageEntry : imageKeyToTopicMap.entrySet())
          {
             int imageKey = imageEntry.getKey();
             ROS2Topic<? extends Packet<?>> imageTopic = imageEntry.getValue();
 
+            // Get the image to publish
             RawImage imageToPublish = imageSensor.getImage(imageKey);
+
+            // Skip if the image is null
+            if (imageToPublish == null)
+               continue;
+
+            // Skip if this image was already published
+            if (lastPublishedSequenceNumbers.containsKey(imageKey) && imageToPublish.getSequenceNumber() == lastPublishedSequenceNumbers.get(imageKey))
+               continue;
```

Comment:
It doesn't make sense that publishing the same image twice would cause problems? We get some image (doesn't matter what it is or what it contains), and that gets handed to the streaming stuff. Then the streaming stuff publish's that image. Does it have knowledge of information that's contained in the image? This seems like a good test case

Reply:
The streaming stuff uses the image's acquisition time to give the image a timestamp. So I'm guessing it gets confused when it tries to send a frame with the same timestamp as the previous

Reply:
So does that mean something is wrong with the `ROS2SRTSensorStreamer` class? Or maybe at a lower level, I haven't looked at that class much yet

Reply:
Yeah, to be honest I don't know where it's "wrong." On one hand I'm tempted to say you shouldn't stream the same image twice using the `ROS2SRTSensorStreamer` (because it's like telling the streamer "here's a new image to send" but giving it an old image). On the other hand you could say the `FFmpegEncoder` should know not to send frames that it's sent before. Perhaps even LibAV should handle that problem internally.

I'm leaning towards adding a check in the `FFmpegEncoder`. Should be pretty simple to do.